design/baremetal-bootstrap.md

```diff
@@ -0,0 +1,116 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Title
+
+baremetal-bootstrap
+
+## Status
+
+provisional
+
+## Table of Contents
+
+## Summary
+
+In a multicluster scenario using
+[Hive](https://github.com/openshift/hive), there's no immediately
+obvious place to run the bootstrap virtual machine for baremetal IPI.
+Currently the user installing the cluster provides a provisioning host
+to run installer which launches the bootstrap VM locally. This may be a
+laptop. With some small modifications, we would be able to support a
+remote libvirt host that a Hive-launched installer could connect to,
+however this host would already need to exist.
+
+Alternatively, the provisioning cluster could host a libvirt VM using
+Kubevirt, or a more lightweight privileged container running libvirt.
+
+Another approach would be to leverage the baremetal-operator running in
+the Hive cluster to provision a *baremetal bootstrap* host, which this
+design document discusses.
+
+## Motivation
+
+Once there's at least one OpenShift cluster, it should be possible to
+use that cluster to bootstrap the next one in a more automated way.
+
+We would like to avoid any significant of prework required for a user to
+deploy additional clusters.
+
+### Goals
+
+ - With the appropriate hardware in place on the right networks, Hive
+   should be able to provision a new cluster with a minimal amount of
+   human interaction.
+
+### Non-Goals
+
+ - Any other deployment mechanism than a baremetal bootstrap host
+  (remote libvirt, etc.)
+
+## Proposal
+
+### Prerequisites
+
+This design requires:
+
+ - An existing cluster with both Hive and baremetal-operator
+ - 4 baremetal servers
+ - At least two different provisioning networks
+
+### Risks and Mitigations
+
+## Design Details
+
+The baremetal IPI install config will need updating to take BMC
+credentials for a hardware bootstrap host, as well as credentials to
+talk to the OpenShift cluster using Hive. The bootstrap host needs to be
+present on two provisioning networks. The one that provisions it from
+the cluster running Hive, and the one it will use to provision the
+masters for the new cluster.
+
+![Network diagram](images/two-networks.png)
+
+As the installer assumes deployment is orchestrated by terraform, we
+will need a terraform provider that can create a BareMetalHost record in
+OpenShift for provisioning the bootstrap host.  There is a [kubernetes
+provider](https://www.terraform.io/docs/providers/kubernetes/index.html),
```

Comment:
I wonder if this one would be useful? Seems to support arbitrary yaml manifests, including CRDs. But it's not exactly an active project. I found it linked from terraform's [community index](https://www.terraform.io/docs/providers/type/community-index.html).

https://gavinbunney.github.io/terraform-provider-kubectl/docs/provider.html

Reply:
Thanks! This doesn't make me feel to confident though: "This is a fork (of a fork!) of the original provider." If it doesn't suit our needs, writing our own isn't a big undertaking, the Ironic provider itself was pretty easy and is generally low maintenance. Something that sends some yaml to kubernetes should be easy (famous last words).

Reply:
I added a link to this provider in the doc

Reply:
Yeah, agreed it shouldn't be too bad. I [did an exercise](https://github.com/mhrivnak/kcreate/blob/master/kcreate.go) some time back just to see how hard it would be to re-implement `kubectl create -f`, and it turned out to be quite simple.